File: restclient/EMClient.py
import requests
from requests.exceptions import RequestException
import json
import os
import sys
import platform
import re
from datetime import datetime

# Constants
BASE_URL = "http://172.32.0.11:32768/api"
GET_COMMANDS_URL = f"{BASE_URL}/get"
CMD_COMMANDS_URL = f"{BASE_URL}/cmd"
COMMANDS_PER_PAGE = 5

def clear_screen():
    """Clears the terminal screen for a fresh menu display."""
    os.system('cls' if platform.system() == 'Windows' else 'clear')

def fetch_commands():
    """Fetches the list of commands from the API.

    Returns:
        dict: The full response with apps and commands.
    """
    try:
        response = requests.get(GET_COMMANDS_URL)
        response.raise_for_status()
        return response.json()
    except requests.RequestException as e:
        print(f"Error fetching commands: {e}")
        return None

def fetch_apps(data):
    """Extracts the list of apps from the response.

    Args:
        data (dict): The full API response.

    Returns:
        list: A list of app dictionaries.
    """
    return data.get('apps', [])

def select_app(apps):
    """Displays available apps and prompts user to select one.

    Args:
        apps (list): List of app dictionaries.

    Returns:
        dict: The selected app object, or None if user cancels.
    """
    clear_screen()
    print("\n" + "=" * 20 + " Available Apps " + "=" * 20)
    
    if not apps:
        print("No apps available.")
        return None

    for idx, app in enumerate(apps, 1):
        description = app.get('description', 'No description available.')
        app_id = app.get('id', 'No ID')
        print(f"{idx}: {description} ({app_id})")

    print(f"{len(apps) + 1}: Exit")

    while True:
        try:
            selection = int(input("\nSelect an app by number: "))
            
            if selection == len(apps) + 1:
                print("Exiting the program.")
                exit(0)
            
            if 1 <= selection <= len(apps):
                return apps[selection - 1]
            else:
                print("Invalid selection, please try again.")
        except ValueError:
            print("Invalid input, please enter a number.")

def display_menu(commands, level=0, page=1):
    """Displays the command menu with pagination.

    Args:
        commands (list): List of command dictionaries.
        level (int): The current menu level.
        page (int): The current page number.

    Returns:
        bool: True if commands were displayed successfully, else False.
    """
    start_index = (page - 1) * COMMANDS_PER_PAGE
    end_index = start_index + COMMANDS_PER_PAGE
    paginated_commands = commands[start_index:end_index]

    clear_screen()
    print("\n" + ("=" * (level + 1)) + " Menu " + ("=" * (level + 1)))

    if not paginated_commands:
        print("No more commands to display.")
        return False

    for idx, command in enumerate(paginated_commands):
        description = command.get('description', 'No description available.')
        command_id = command.get('id', 'No ID')
        print(f"{start_index + idx + 1}: {description} ({command_id})")

    if level == 0:
        print(f"{len(paginated_commands) + 1}: Exit")
    else:
        print(f"{len(paginated_commands) + 1}: Back to previous menu")

    return True

def is_valid(input_value, arg_type):
    """Validates user input based on field type.

    Args:
        input_value (str): The value to validate.
        arg_type (str): The type of the argument.

    Returns:
        bool: True if input is valid, else False.
    """
    if '@' in arg_type:
        arg_type = arg_type.split('@')[-1]

    if arg_type == "str":
        if input_value.strip() == "":
            print("Hint: Input should be a non-empty string.")
            return False
        return True

    if arg_type == "int" and re.match(r"-?\d+", input_value):
        return True
    elif arg_type == "unsigned" and re.match(r"\d+", input_value):
        return True
    elif arg_type == "date":
        try:
            datetime.strptime(input_value, "%Y-%m-%d")
            return True
        except ValueError:
            print("Hint: Input should be a valid date in the format YYYY-MM-DD.")
            return False
    elif arg_type == "time":
        try:
            datetime.strptime(input_value, "%H:%M")
            return True
        except ValueError:
            print("Hint: Input should be a valid time in HH:mm format.")
            return False
    elif arg_type == "duration" and re.match(r"^\d+h \d+m$", input_value):
        return True

    print(f"Hint: Unknown argument type: {arg_type}")
    return False

def get_action_type(selected_command):
    """Extracts the action type from the command.
    
    Args:
        selected_command (dict): Dictionary of the selected command.
    
    Returns:
        str: The action type (e.g., 'add' from 'event.add', or 'init' if no dot exists).
    """
    action = selected_command['action']
    if '.' in action:
        return action.split('.')[1]
    return 'init'


def get_default_value(field_info, action):
    """Gets the default value for a field based on the action.
    
    Args:
        field_info (dict): Field information including defaultValue.
        action (str): The current action type.
    
    Returns:
        str: The default value for the action, or empty string if not found.
    """
    default_value = field_info.get('defaultvalue', {})
    
    # If defaultValue is a string (old format), return it
    if isinstance(default_value, str):
        return default_value
    
    # If defaultValue is a dict (new format), get value for the action
    if isinstance(default_value, dict):
        return default_value.get(action, '')
    
    return ''


def is_field_supported(field_info, action):
    """Checks if the field supports the current action.
    
    Args:
        field_info (dict): Field information including supports list.
        action (str): The current action type.
    
    Returns:
        bool: True if the field supports the action or has no supports restriction.
    """
    supports = field_info.get('supports', [])
    
    # If no supports list, field is supported for all actions
    if not supports:
        return True
    
    # Check if current action is in the supports list
    return action in supports


def get_user_input(field_info, action):
    """Prompts user for input based on field information and action.

    Args:
        field_info (dict): Field information including name, type, etc.
        action (str): The current action type.

    Returns:
        str: User input value or default value.
    """
    field_name = field_info['field']
    description = field_info.get('description', field_name)
    # Check if referencedfield exists and has a type, otherwise use the main type
    referenced_field = field_info.get('referencedfield')
    field_type = referenced_field.get('type') if referenced_field else field_info.get('type')

    mandatory = field_info.get('mandatory', False)
    modifier = field_info.get('modifier')
    default_value = get_default_value(field_info, action)

    if modifier == "auto":
        return default_value

    # Add * to description if mandatory
    prompt_description = f"{description}*" if mandatory else description

    while True:
        user_input = input(f"Enter {prompt_description} ({field_type}): ")

        if mandatory and user_input.strip() == "":
            print(f"{description} is mandatory. Please provide a value.")
            continue

        if not mandatory and user_input.strip() == "":
            return default_value

        if is_valid(user_input, field_type):
            return int(user_input) if field_type == "int" else user_input

        print("Invalid input. Please try again.")

def create_payload(args_info, action):
    """Creates a JSON payload from user inputs.

    Args:
        args_info (dict): Dictionary of field information.
        action (str): The current action type.

    Returns:
        dict: Dictionary representing the payload.
    """
    payload = {}

    for field_key, field_info in args_info.items():
        # Check if field supports current action
        if not is_field_supported(field_info, action):
            continue
        
        user_input = get_user_input(field_info, action)
        payload[field_key] = user_input

    return payload

def run_command(selected_command, root_identifier, app_id):
    """Executes the command with the provided identifier.

    Args:
        selected_command (dict): Dictionary of the selected command.
        root_identifier (str): Identifier from the root.
        app_id (str): The ID of the selected app.
    """
    action = get_action_type(selected_command)
    
    payload = {
        "identifier": f"{root_identifier}",
        "commands": [{
            "args": {},
            "data": {},
            "id": selected_command['action']
        }]
    }

    if 'args' in selected_command:
        for field_key, field_info in selected_command['args'].items():
            # Check if field supports current action
            if not is_field_supported(field_info, action):
                continue
            
            user_input = get_user_input(field_info, action)
            if user_input is not None:
                payload['commands'][0]['data'][field_key] = user_input

    try:
        print("Payload Sent:")
        print(json.dumps(payload, indent=2))
        response = requests.post(CMD_COMMANDS_URL, json=payload)
        response.raise_for_status()
        print("Response from the server:")
        print(json.dumps(response.json(), indent=2))
    except requests.RequestException as e:
        print(f"Error sending command: {e}")

def handle_selection(commands, app_id, level=0):
    """Handles user selection from the menu.

    Args:
        commands (list): List of command dictionaries.
        app_id (str): The ID of the selected app.
        level (int): The current menu level.
    """
    page = 1

    while True:
        displayed = display_menu(commands, level, page)

        if not displayed:
            break

        selection = input("Select a command by number: ")

        try:
            selection = int(selection)
        except ValueError:
            print("Invalid selection, please enter a number.")
            continue

        if level == 0 and selection == len(commands) + 1:
            print("Exiting the program.")
            exit(0)
        elif selection == len(commands) + 1:
            page += 1
            continue
        elif selection == len(commands) + 2:
            return

        if 1 <= selection <= len(commands):
            selected_command = commands[selection - 1]
            if 'commands' in selected_command:
                handle_selection(selected_command['commands'], app_id, level + 1)
            else:
                root_identifier = selected_command.get('id', 'root').split('.')[0] + '@' + app_id
                run_command(selected_command, root_identifier, app_id)
                input("Press Enter to continue...")
        else:
            print("Invalid selection, please try again.")

def main():
    """Main function to execute the command-line interface."""
    data = fetch_commands()
    if data is None:
        return

    apps = fetch_apps(data)
    if not apps:
        print("No apps available.")
        return

    selected_app = select_app(apps)
    if selected_app is None:
        return

    app_id = selected_app.get('id')
    app_commands = selected_app.get('commands', [])
    if not app_commands:
        print(f"No commands available for app: {app_id}")
        return

    clear_screen()
    print(f"\nLoaded app: {selected_app.get('description', app_id)}")
    print("Starting command menu...\n")
    input("Press Enter to continue...")

    while True:
        clear_screen()
        display_menu(app_commands)
        handle_selection(app_commands, app_id)

def check_service_availability():
    try:
        response = requests.get(GET_COMMANDS_URL, timeout=5)
        response.raise_for_status()
        print("Service is available.")
        return True
    except RequestException as e:
        print(f"Service is not available")
        return False

if __name__ == "__main__":
    if check_service_availability():
        main()
    else:
        print("Exiting due to service unavailability.")
        exit(1)

File: restclient/EMClient.js
// Constants
const BASE_URL = "http://172.32.0.11:32768/api";
const GET_COMMANDS_URL = `${BASE_URL}/get`;
const CMD_COMMANDS_URL = `${BASE_URL}/cmd`;
const COMMANDS_PER_PAGE = 5;

const readline = require("readline");

// Create a single readline interface that persists throughout the application
const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

// Utility function to prompt user
function prompt(question) {
  return new Promise((resolve) => {
    rl.question(question, (answer) => {
      resolve(answer);
    });
  });
}

// Utility Functions
function clearScreen() {
  /**
   * Clears the console
   */
  console.clear();
}

async function fetchJson() {
  /**
   * Fetches the entire JSON from the API.
   * @returns {Promise<Object|null>} The JSON object or null on error
   */
  try {
    const response = await fetch(GET_COMMANDS_URL);
    if (!response.ok) {
      throw new Error(`HTTP error! status: ${response.status}`);
    }
    const data = await response.json();
    return data;
  } catch (error) {
    console.error(`Error fetching JSON: ${error.message}`);
    return null;
  }
}

function displayAppMenu(apps) {
  /**
   * Displays the app selection menu.
   * @param {Array} apps - List of app objects
   */
  clearScreen();
  console.log("\n" + "=".repeat(5) + " Select an App " + "=".repeat(5));

  apps.forEach((app, idx) => {
    const appId = app.id || "Unknown";
    const description = app.description || "No description available.";
    console.log(`${idx + 1}: ${appId} - ${description}`);
  });

  console.log(`${apps.length + 1}: Exit`);
}

function displayCommandMenu(appId, commands, page = 1) {
  /**
   * Displays the command menu with pagination.
   * @param {string} appId - The selected app ID
   * @param {Array} commands - List of command objects
   * @param {number} page - Current page number
   * @returns {Object} Object containing display info and paginated commands
   */
  const startIndex = (page - 1) * COMMANDS_PER_PAGE;
  const endIndex = startIndex + COMMANDS_PER_PAGE;
  const paginatedCommands = commands.slice(startIndex, endIndex);

  clearScreen();
  console.log("\n" + "=".repeat(5) + ` Commands for: ${appId} ` + "=".repeat(5));

  if (paginatedCommands.length === 0) {
    console.log("No more commands to display.");
    return { 
      displayed: false, 
      paginatedCommands: [], 
      totalCommands: commands.length 
    };
  }

  paginatedCommands.forEach((command, idx) => {
    const description = command.description || "No description available.";
    const commandId = command.id || "No ID";
    console.log(`${startIndex + idx + 1}: ${description} (${commandId})`);
  });

  const nextOptionNumber = paginatedCommands.length + 1;
  console.log(`${nextOptionNumber}: Back`);
  console.log(`${nextOptionNumber + 1}: Exit`);

  return {
    displayed: true,
    paginatedCommands,
    totalCommands: commands.length,
    currentPage: page,
    backOption: nextOptionNumber,
    exitOption: nextOptionNumber + 1,
  };
}

function isValid(inputValue, argType) {
  /**
   * Validates user input based on field type.
   * @param {string} inputValue - The value to validate
   * @param {string} argType - The type of the argument
   * @returns {boolean} True if input is valid
   */
  let type = argType;
  if (argType.includes("@")) {
    type = argType.split("@").pop();
  }

  if (type === "str") {
    if (inputValue.trim() === "") {
      console.log("Hint: Input should be a non-empty string.");
      return false;
    }
    return true;
  }

  if (type === "int" && /^-?\d+$/.test(inputValue)) {
    return true;
  } else if (type === "unsigned" && /^\d+$/.test(inputValue)) {
    return true;
  } else if (type === "date") {
    const dateRegex = /^\d{4}-\d{2}-\d{2}$/;
    if (!dateRegex.test(inputValue)) {
      console.log("Hint: Input should be a valid date in the format YYYY-MM-DD.");
      return false;
    }
    const date = new Date(inputValue);
    return !isNaN(date.getTime());
  } else if (type === "time") {
    const timeRegex = /^([0-1][0-9]|2[0-3]):[0-5][0-9]$/;
    if (!timeRegex.test(inputValue)) {
      console.log("Hint: Input should be a valid time in HH:mm format.");
      return false;
    }
    return true;
  } else if (type === "duration" && /^\d+h \d+m$/.test(inputValue)) {
    return true;
  }

  console.log(`Hint: Unknown argument type: ${type}`);
  return false;
}

function getActionType(selectedCommand) {
  /**
   * Extracts the action type from the command.
   * @param {Object} selectedCommand - The selected command object
   * @returns {string} The action type (e.g., 'add' from 'event.add', or 'init' if no dot exists)
   */
  const action = selectedCommand.action;
  if (action.includes(".")) {
    return action.split(".")[1];
  }
  return "init";
}

function getDefaultValue(fieldInfo, action) {
  /**
   * Gets the default value for a field based on the action.
   * @param {Object} fieldInfo - Field information including defaultValue
   * @param {string} action - The current action type
   * @returns {string} The default value for the action, or empty string if not found
   */
  const defaultValue = fieldInfo.defaultvalue || {};

  // If defaultValue is a string (old format), return it
  if (typeof defaultValue === "string") {
    return defaultValue;
  }

  // If defaultValue is an object (new format), get value for the action
  if (typeof defaultValue === "object") {
    return defaultValue[action] || "";
  }

  return "";
}

function isFieldSupported(fieldInfo, action) {
  /**
   * Checks if the field supports the current action.
   * @param {Object} fieldInfo - Field information including supports list
   * @param {string} action - The current action type
   * @returns {boolean} True if the field supports the action or has no supports restriction
   */
  const supports = fieldInfo.supports || [];

  // If no supports list, field is supported for all actions
  if (supports.length === 0) {
    return true;
  }

  // Check if current action is in the supports list
  return supports.includes(action);
}

async function getUserInput(fieldInfo, action) {
  /**
   * Prompts user for input based on field information and action.
   * @param {Object} fieldInfo - Field information including name, type, etc.
   * @param {string} action - The current action type
   * @returns {Promise<string|number>} User input value or default value
   */
  const fieldName = fieldInfo.field;
  const description = fieldInfo.description || fieldName;
  // Check if referencedfield exists and has a type, otherwise use the main type
  const fieldType = fieldInfo.referencedfield?.type || fieldInfo.type;
  const mandatory = fieldInfo.mandatory || false;
  const modifier = fieldInfo.modifier;
  const defaultValue = getDefaultValue(fieldInfo, action);

  if (modifier === "auto") {
    return defaultValue;
  }

  // Add * to description if mandatory
  const promptDescription = mandatory ? `${description}*` : description;

  while (true) {
    const userInput = await prompt(`Enter ${promptDescription} (${fieldType}): `);

    if (mandatory && userInput.trim() === "") {
      console.log(`${description} is mandatory. Please provide a value.`);
      continue;
    }

    if (!mandatory && userInput.trim() === "") {
      return defaultValue;
    }

    if (isValid(userInput, fieldType)) {
      return fieldType === "int" ? parseInt(userInput) : userInput;
    }

    console.log("Invalid input. Please try again.");
  }
}

async function createPayload(argsInfo, action) {
  /**
   * Creates a JSON payload from user inputs.
   * @param {Object} argsInfo - Dictionary of field information
   * @param {string} action - The current action type
   * @returns {Promise<Object>} Dictionary representing the payload
   */
  const payload = {};

  for (const [fieldKey, fieldInfo] of Object.entries(argsInfo)) {
    // Check if field supports current action
    if (!isFieldSupported(fieldInfo, action)) {
      continue;
    }

    // Skip fields with modifier "auto"
    if (fieldInfo.modifier === "auto") {
      continue;
    }

    const userInput = await getUserInput(fieldInfo, action);
    payload[fieldKey] = userInput;
  }

  return payload;
}

async function runCommand(selectedCommand, appId) {
  /**
   * Executes the command with the provided app identifier.
   * @param {Object} selectedCommand - The selected command object
   * @param {string} appId - The app ID
   */
  const action = getActionType(selectedCommand);

  const payload = {
    identifier: appId,
    commands: [
      {
        args: {},
        data: {},
        id: selectedCommand.action,
      },
    ],
  };

  if (selectedCommand.args) {
    for (const [fieldKey, fieldInfo] of Object.entries(selectedCommand.args)) {
      // Check if field supports current action
      if (!isFieldSupported(fieldInfo, action)) {
        continue;
      }

      // Skip fields with modifier "auto"
      if (fieldInfo.modifier === "auto") {
        continue;
      }

      const userInput = await getUserInput(fieldInfo, action);
      //payload.commands[0].args[fieldKey] = fieldInfo;
      if (userInput !== null) {
        payload.commands[0].data[fieldKey] = userInput;
      }
    }
  }

  try {
    const response = await fetch(CMD_COMMANDS_URL, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify(payload),
    });

    if (!response.ok) {
      throw new Error(`HTTP error! status: ${response.status}`);
    }

    console.log("Payload Sent:");
    console.log(JSON.stringify(payload, null, 2));
    const responseData = await response.json();
    console.log("Response from the server:");
    console.log(JSON.stringify(responseData, null, 2));
  } catch (error) {
    console.error(`Error sending command: ${error.message}`);
  }
}

async function handleAppSelection(json) {
  /**
   * Handles app selection from the menu.
   * @param {Object} json - The full JSON object containing apps
   */
  const apps = json.apps || [];

  if (apps.length === 0) {
    console.log("No apps available.");
    return;
  }

  while (true) {
    displayAppMenu(apps);

    const selection = await prompt("Select an app by number: ");
    let selectedNum;

    try {
      selectedNum = parseInt(selection);
    } catch (error) {
      console.log("Invalid selection, please enter a number.");
      continue;
    }

    // Handle "Exit" option
    if (selectedNum === apps.length + 1) {
      console.log("Exiting the program.");
      rl.close();
      process.exit(0);
    }

    // Handle app selection
    if (selectedNum >= 1 && selectedNum <= apps.length) {
      const selectedApp = apps[selectedNum - 1];
      const appId = selectedApp.id;
      const commands = selectedApp.commands || [];

      if (commands.length === 0) {
        console.log("No commands available for this app.");
        await prompt("Press Enter to continue...");
        continue;
      }

      await handleCommandSelection(appId, commands);
    } else {
      console.log("Invalid selection, please try again.");
    }
  }
}

async function handleCommandSelection(appId, commands, level = 0) {
  /**
   * Handles command selection from the menu.
   * @param {string} appId - The selected app ID
   * @param {Array} commands - List of command objects
   * @param {number} level - Current menu level (for nested commands)
   */
  let page = 1;

  while (true) {
    const menuInfo = displayCommandMenu(appId, commands, page);

    if (!menuInfo.displayed) {
      break;
    }

    const selection = await prompt("Select a command by number: ");
    let selectedNum;

    try {
      selectedNum = parseInt(selection);
    } catch (error) {
      console.log("Invalid selection, please enter a number.");
      continue;
    }

    const paginatedCommands = menuInfo.paginatedCommands;
    const backOption = menuInfo.backOption;
    const exitOption = menuInfo.exitOption;

    // Handle "Back" option
    if (selectedNum === backOption) {
      return;
    }

    // Handle "Exit" option
    if (selectedNum === exitOption) {
      console.log("Exiting the program.");
      rl.close();
      process.exit(0);
    }

    // Handle command selection
    if (selectedNum >= 1 && selectedNum <= paginatedCommands.length) {
      const selectedCommand = paginatedCommands[selectedNum - 1];

      if (selectedCommand.commands && selectedCommand.commands.length > 0) {
        // Nested menu
        await handleCommandSelection(selectedCommand.id || appId, selectedCommand.commands, level + 1);
      } else {
        // Execute command
        await runCommand(selectedCommand, appId);
        await prompt("Press Enter to continue...");
      }
    } else {
      console.log("Invalid selection, please try again.");
    }
  }
}

async function main() {
  /**
   * Main function to execute the command-line interface.
   */
  const json = await fetchJson();
  if (json === null) {
    return;
  }

  // Check if apps exist in JSON
  if (!json.apps || json.apps.length === 0) {
    console.log("No apps available in the JSON response.");
    rl.close();
    process.exit(1);
  }

  await handleAppSelection(json);
}

async function checkServiceAvailability() {
  /**
   * Checks if the REST service is available.
   * @returns {Promise<boolean>} True if service is available
   */
  try {
    const response = await fetch(GET_COMMANDS_URL, { timeout: 5000 });
    if (!response.ok) {
      throw new Error(`HTTP error! status: ${response.status}`);
    }
    console.log("Service is available.");
    return true;
  } catch (error) {
    console.log("Service is not available");
    return false;
  }
}

// Main execution
(async () => {
  if (await checkServiceAvailability()) {
    await main();
  } else {
    console.log("Exiting due to service unavailability.");
    rl.close();
    process.exit(1);
  }
})();
